### PR TITLE
Implement safe plugin uninstallation

### PR DESCRIPTION
**Pull request by colin-grant-work**
#### What it does

This PR implements several things:
1. Deployment of folders to a temp dir, with a CLI option to prevent that copying.
    - Install some unzipped plugins in your `~/.theia/extensions` directory
    - Observe that during startup, messages are logged about copying those plugins.
    - Use the `--uncompressed-plugins-in-place` flag in the CLI
    - Observe that you don't get those startup messages.
2. Uninstallation without undeployment
    - Install some user plugins
    - Open more than one instance of Theia on the same backend
    - Open the plugin view in one instance
    - Uninstall one of your user plugins
    - Observe that the plugin's features continue to work in all instances.
    - Observe that in all instances targeting that backend, that plugin now shows a button saying 'Reload Required'
    - Click that button in one or more instances.
    - Observe that when the instances reload, the plugin is not loaded.
3. Multiple version handling
    - Follow the same steps as (2) up to 'Uninstall one of your user plugins'
    - After uninstalling the plugin, use 'Install from VSIX' to install a different version of that plugin
    - Observe that you still have the 'reload required' button
    - Click it
    - When that instance reloads, it should have access to the new version of the plugin.
4. Reimplementation of #10951
    - Follow testing steps from #10951
    - Follow testing steps from #11057 with the caveat that the file should appear in your home directory, but not be decompressed.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by colin-grant-work** [comment]
@msujew, thanks for taking a look!

> However, I noticed something which didn't work as described in the testing steps. "Installing" an extension by unzipping it into `~/.theia` and starting the app did not copy the extension into the `/tmp/vscode-unpacked` directory (with and without the `--uncompressed-plugins-in-place` flag). It did work for zipped (`.vsix`) extensions as expected. Is that something I got incorrect?

It should be copying all directories from the user folder. You said you unzipped into the `~/.theia/extensions` (or just `~/.theia`?) directory; what was your exact procedure?
  
> Aside from that I noticed that every uninstalled extension now requires to reload before disappearing completely. While this is definitely an improvement to the way it worked previously, vscode seems to be able to decide on a case-by-case basis whether uninstalling an extension requires an app reload. But this is probably something for another PR :)

I have an idea for this, but it's harder to guarantee on the backend than the frontend, so I've left it for later :-).

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 1]
```diff
@@ -795,6 +799,7 @@ export interface PluginMetadata {
     host: string;
     model: PluginModel;
     lifecycle: PluginLifecycle;
+    outOfSynch: boolean;
```
[the remark this answers is not in the document]
Comment (colin-grant-work) [reply packages/plugin-ext/src/common/plugin-protocol.ts:802]: Not a typo, but probably shouldn't be there anyway. In Greek, the 'ch' is one letter, so I don't like that in English we've decided to remove the 'h' in some cases. But it is indeed convention, and one should follow convention :-).

**Comment by alvsan09** [comment]
> It should be copying all directories from the user folder. You said you unzipped into the `~/.theia/extensions` (or just `~/.theia`?) directory; what was your exact procedure?

There is a typo on the instructions at the top, it should be `~/.theia/extensions` as you just mentioned rather than ``~/.theia/plugins`

**Comment by colin-grant-work** [comment]
> > It should be copying all directories from the user folder. You said you unzipped into the `~/.theia/extensions` (or just `~/.theia`?) directory; what was your exact procedure?
> 
> @colin-grant-work There is a typo on the instructions at the top, it should be `~/.theia/extensions` as you just mentioned rather than ``~/.theia/plugins`

Good catch. I've also fixed this behavior. @msujew, it looks like during a rebase, the actual call to copy VSCode plugins got deleted; I've brought it back in [this commit](https://github.com/eclipse-theia/theia/pull/11084/commits/7a596997f37593647666e43bf4de02872da1be11)

File: packages/plugin-ext/src/main/node/plugin-cli-contribution.ts [hunk 2]
```diff
@@ -70,4 +79,7 @@ export class PluginCliContribution implements CliContribution {
         return this._maxSessionLogsFolders;
     }
 
+    copyUncomprossedPlugins(): boolean {
```
Comment (alvsan09) [inline packages/plugin-ext/src/main/node/plugin-cli-contribution.ts:82]: Typo 'Uncomprossed'

**Review by alvsan09 — COMMENTED** [review]
I have tested the following scenarios and they work fine! :+1: 
1. Tested deployment of folders to a 'tmp' directory
2. Tested loading of plugin folders in place via `--uncompressed-plugins-in-place` flag
3. Uninstallation without undeployment for .vsix
4. Uninstallation without undeployment for uncompressed folder extensions
5. Uninstalling and installing a different version

I am not too familiar with the code base, so I will let more experienced committers in the area to cover it.
A found a typo and mentioned it as an inline comment.

**Comment by colin-grant-work** [comment]
@tsmaeder, it might be good to have a Red Hat perspective on this, since it makes substantial changes to the way we pass around data about plugins - changing and making explicit expectations about whether versions are included as part of identifiers - and adds some state that we didn't have before. Want to make sure that we don't blow up Che, if they're still updating the version of Theia they offer.

**Review by vince-fugnitto — APPROVED** [review]
The changes work well for me :+1: 

I confirmed that:
- [x] extensions can be searched.
- [x] extensions can be installed.
- [x] extensions can be uninstalled.
- [x] uninstalled extensions require a `reload` before they are removed.
- [x] upon reloading the application extensions can be re-installed.
- [x] if an extensions is uninstalled, multiple clients will be notified and display `reload required`.
- [x] if an extension is uninstalled, and installed through `install from vsix` the proper extension versions is used on reload.
- [x]  deployment of extensions works well.

**Review by msujew — APPROVED** [review]
Reconfirming my approval. With the latest changes putting an unzipped extension into the `~/.theia/plugins` folder copies it to the `/tmp/vscode-copied` directory and activates it correctly. The behavior for the flag also works as expected 👍

File: packages/core/src/common/promise-util.ts [hunk 3]
```diff
@@ -24,7 +24,7 @@ import { CancellationToken, CancellationError, cancelled } from './cancellation'
  */
 export class Deferred<T = void> {
     state: 'resolved' | 'rejected' | 'unresolved' = 'unresolved';
-    resolve: (value: T) => void;
+    resolve: (value: T | PromiseLike<T>) => void;
```
[the remark this answers is not in the document]
Comment (paul-marechal) [reply packages/core/src/common/promise-util.ts:27]: On the other hand I could be misunderstanding the role of the `state` property?

Is it meant to represent when `resolve` as been called, or when the underlying promise resolves? Doesn't look obvious.

**Comment by colin-grant-work** [comment]
Thanks everyone who reviewed!

**Comment by tsmaeder** [comment]
@colin-grant-work would have loved to chime in, but no time, unfortunately 🤷

**Pull request merged**